Sources/NIOOpenSSL/SSLPKCS12Bundle.swift

```diff
@@ -0,0 +1,218 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOOpenSSL
+import NIO
+
+
+/// A container of a single PKCS#12 bundle.
+///
+/// PKCS#12 is a specification that defines an archive format for storing multiple
+/// cryptographic objects together in one file. Its most common usage, and the one
+/// that SwiftNIO is most interested in, is its use to bundle one or more X.509
+/// certificates (`OpenSSLCertificate`) together with an associated private key
+/// (`OpenSSLPrivateKey`).
+///
+/// ### Working with TLSConfiguration
+///
+/// In many cases users will want to configure a `TLSConfiguration` with the data
+/// from a PKCS#12 bundle. This object assists in unpacking that bundle into its
+/// associated pieces.
+///
+/// If you have a PKCS12 bundle, you configure a `TLSConfiguration` like this:
+///
+///    let p12Bundle = OpenSSLPKCS12Bundle(file: pathToMyP12)
+///    let config = TLSConfiguration.forServer(certificateChain: p12Bundle.certificateChain,
+///                                            privateKey: p12Bundle.privateKey)
+///
+/// The created `TLSConfiguration` can then be safely used for your endpoint.
+public struct OpenSSLPKCS12Bundle {
+    public let certificateChain: [OpenSSLCertificate]
+    public let privateKey: OpenSSLPrivateKey
+
+    private init<Bytes: Collection>(ref: UnsafeMutablePointer<PKCS12>, passphrase: Bytes?) throws where Bytes.Element == UInt8 {
+        var pkey: UnsafeMutablePointer<EVP_PKEY>? = nil
+        var cert: UnsafeMutablePointer<X509>? = nil
+        var caCerts: UnsafeMutablePointer<stack_st_X509>? = nil
+
+        let rc = try passphrase.withSecureBytes { passphrase in
+            PKCS12_parse(ref, passphrase, &pkey, &cert, &caCerts)
+        }
+        guard rc == 1 else {
+            throw OpenSSLError.unknownError(.init())
+        }
+
+        // Successfully parsed, let's unpack. The key and cert are mandatory,
+        // the ca stack is not.
+        guard let actualCert = cert, let actualKey = pkey else {
+            // Free the pointers that we have.
+            X509_free(cert)
+            EVP_PKEY_free(pkey)
+            CNIOOpenSSL_sk_X509_free(caCerts)
+            throw NIOOpenSSLError.unableToAllocateOpenSSLObject
+        }
+
+        let certStackSize = caCerts.map(CNIOOpenSSL_sk_X509_num) ?? 0
+        var certs = [OpenSSLCertificate]()
+        certs.reserveCapacity(Int(certStackSize) + 1)
+        certs.append(OpenSSLCertificate.fromUnsafePointer(pointer: actualCert))
```

Comment:
I'm assuming the `fromUnsafePointer(pointer: ...)` functions are consuming a reference? If yes, could we file an issue to rewrite them to something that doesn't consume the reference and copies the contents instead. If OpenSSL doesn't have copying functions for that then the label needs to become better and needs to read `fromUnsafePointer(transferringOwnership: actualCert)` etc...

Reply:
and we need a comment at each invocation site that they're non-standard and just take ownership of the passed pointer.